feat:  Add `RedisStorageClient`

Mantisus opened the pull request:
### Description

- TODO

### Testing

- TODO

Mantisus commented:
In `RedisRequestQueueClient`, I used a [Bloom filter](https://en.wikipedia.org/wiki/Bloom_filter) in [`Redis`](https://redis.io/docs/latest/develop/data-types/probabilistic/bloom-filter/) for deduplication and tracking handled requests. This approach differs from what we used in other clients (using `set`). My main motivation was that Redis is an in-memory database, and memory consumption can be critical in some cases.

Since a Bloom filter is a probabilistic data structure, the final data structure size is affected by the error probability; I used `1e-7`. This means that with a probability of `1e-7`, we may get a false positive when checking the filter. In our case, this translates to a probability of skipping a request: with probability `1e-7`, a request that wasn't added to the queue will be considered as already added. Similarly, with probability `1e-7`, a request that hasn't been handled yet will be considered as already processed.

Memory consumption for records in the format 'https://crawlee.dev/{i}' (record size doesn't matter for Bloom filters):

**Redis Bloom filter:**
- 100,000 - 427 KB
- 1,000,000 - 4 MB  
- 10,000,000 - 42 MB

**Redis set:**
- 100,000 - 6 MB
- 1,000,000 - 61 MB
- 10,000,000 - 662 MB

Discussion about whether it's worth pursuing this approach is welcome!

janbuchar commented:
I haven't read the PR yet, but I did look into bloom filters for request deduplication in the past and what you wrote piqued my interest 🙂 I am a little worried about the chance of dropping a URL completely, even with a super small probability.

Perhaps we should default to a solution that tolerates some percentage of the "opposite" errors and allows a URL to get processed multiple times in rare cases. A fixed size hash table is an example of such data structure. I don't know if anything more sophisticated exists.

But maybe I have an irrational fear of probabilistic stuff 🙂

Mantisus commented:
Since we have already added the ability to parameterize queue behavior in the SDK (`single`, `shared`), I updated `RedisStorageClient` with a parameter defining the deduplication strategy `queue_dedup_strategy`. If set to `default`, it will use Set, `bloom` Bloom filters.

Pijukatel commented on src/crawlee/storage_clients/_redis/_request_queue_client.py:
```diff
@@ -0,0 +1,578 @@
+    _MAX_BATCH_FETCH_SIZE = 10
+    """Maximum number of requests to fetch in a single batch operation."""
+
+    _BLOCK_REQUEST_TIME = 300_000  # milliseconds
```
Could you please explain this a bit?

Based on the description, I get the impression that I can't call reclaim_request for 300 s, but that is probably not the case.

Mantisus replied:
This is the time for which the request is blocked in the storage for other clients. After it expires, it will be returned to the queue. This protects against queue freezing if any of the clients crash.

I will improve the documentation.

Pijukatel commented on src/crawlee/storage_clients/_redis/_request_queue_client.py:
```diff
@@ -0,0 +1,578 @@
+            delta_total = len(actually_added)
+
+            for unique_key in new_unique_keys:
+                if unique_key in actually_added:
```
Maybe simplify to
```
  processed_requests.append(
      ProcessedRequest(
          unique_key=unique_key,
          was_already_present=not unique_key in actually_added,
          was_already_handled=False,
      )
  )
```

Pijukatel reviewed (commented):
Just some small comments and questions. I am really looking forward to using this. Good work.

Mantisus commented on src/crawlee/storage_clients/_redis/_client_mixin.py:
```diff
@@ -0,0 +1,297 @@
+            )
+            client = cls(storage_name=internal_name, storage_id=metadata.id, redis=redis, **instance_kwargs)
+            created = await client._create_metadata_and_storage(internal_name, metadata.model_dump())
+            # The client was probably not created due to a race condition. Let's try to open it using the name.
```
Yes, that's why the `with_wait` synchronizer is used. So that when the client encounters a race condition, it waits until the first client finishes creating the storage and reports it.

janbuchar commented on src/crawlee/storage_clients/_redis/_utils.py:
```diff
@@ -0,0 +1,22 @@
+    return await response if isinstance(response, Awaitable) else response
+
+
+def read_lua_script(file_path: Path) -> str:
```
Couldn't this accept just the file name and prepend the path to the `lua_scripts` directory automatically?

Mantisus commented on tests/unit/storage_clients/_redis/test_redis_dataset_client.py:
```diff
@@ -0,0 +1,179 @@
+    name = await await_redis_response(dataset_client.redis.hget('datasets:id_to_name', metadata.id))
+
+    assert name is not None
+    assert (name.decode() if isinstance(name, bytes) else name) == 'test_dataset'
```
Due to the `decode_responses` parameter in the Redis constructor.

janbuchar reviewed (commented):
I think we should reach out to somebody who knows Redis better to review this for us. Or do you feel confident @Pijukatel? 😁

Mantisus commented on src/crawlee/storage_clients/_redis/_client_mixin.py:
```diff
@@ -0,0 +1,297 @@
+        Returns:
+            An instance for the opened or created storage client.
+        """
+        internal_name = name or alias or cls._DEFAULT_NAME
```
for alias in the metadata object `name=None`.

internal_name is used to form the key prefix. For example, in `FileSystemStorageClient`, the folder name is formed.

vdusek commented on docs/guides/storage_clients.mdx:
```diff
@@ -313,6 +319,172 @@ For advanced scenarios, you can configure <ApiLink to="class/SqlStorageClient">`
+%% Dataset Keys
+%% ========================
+
+class Dataset_Keys {
```
Do we follow `PascalCase` combined with `snake_case` here? 🙂

vdusek commented on tests/unit/storages/conftest.py:
```diff
@@ -1,17 +1,40 @@
 
-@pytest.fixture(params=['memory', 'file_system', 'sql'])
-def storage_client(request: pytest.FixtureRequest) -> StorageClient:
+@pytest.fixture(params=['memory', 'file_system', 'sql', 'redis_default', 'redis_bloom'])
```
I'm not sure if it makes sense to have two Redis storage clients here, as that would result in many more test runs covering the same functionality, just due to a different type of deduplication in Redis RQ. Maybe we could test Redis RQ with both the default and bloom deduplication in a more targeted way instead.

Mantisus commented on tests/unit/storage_clients/_redis/test_redis_dataset_client.py:
```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
```
Not many tests could be removed. This is because client tests check interaction with Redis at a lower level than storage tests.

janbuchar commented on docs/examples/playwright_crawler_with_fingerprint_generator.mdx:
um? 🙂

vdusek commented on pyproject.toml:
```diff
@@ -98,6 +99,7 @@ dev = [
     "apify_client", # For e2e tests.
     "build<2.0.0", # For e2e tests.
     "dycw-pytest-only<3.0.0",
+    "fakeredis[probabilistic,json,lua]~=2.32.0",
```
```suggestion
    "fakeredis[probabilistic,json,lua]<3.0.0",
```
